lib/liquid_voting/delegations.ex

```diff
@@ -169,28 +178,88 @@ defmodule LiquidVoting.Delegations do
     proposal_url = Map.get(attrs, :proposal_url)
 
     Delegation
-    |> where([d], d.delegator_id == ^delegator_id)
-    |> where_proposal(proposal_url)
-    |> Repo.one()
+    |> where(delegator_id: ^delegator_id)
+    |> Repo.all()
+    |> resolve_conflicts(delegate_id, proposal_url)
     |> case do
-      # Delegation (of same type) not found, so we build one
-      nil -> %Delegation{}
-      # Delegation (of same type) exists - let's use it
-      delegation -> delegation
+      {:ok, delegations_of_delegator} ->
+        delegations_of_delegator
+        |> find_similar_delegation_or_return_new_struct(proposal_url)
+        |> Delegation.changeset(attrs)
+        |> Repo.insert_or_update()
+
+      {:error, %{message: message, details: details}} ->
+        {:error, %{message: message, details: details}}
     end
-    |> Delegation.changeset(attrs)
-    |> Repo.insert_or_update()
   end
 
-  defp where_proposal(query, _proposal_url = nil),
-    do: query |> where([d], is_nil(d.proposal_url))
+  # Resolves conflicting delegations for same delegator and delegate (2 clauses).
+  #
+  # Used by upsert_delegation/1 (above).
+  #
+  # Clause 1: Matches an attempt to upsert a global delegation.
+  #
+  # Searches list of delegations for the specified delegator for any
+  # proposal-specific delegations to the same delegate in the attributes passed
+  # into upsert_delegation/1. and deletes any such delegations found.
+  #
+  # Clause 2: Matches an attempt to upsert a proposal-specific delegation.
+  #
+  # Searches list of delegations for the specified delegator for global delegation
+  # to the same delegate in the attributes passed into upsert_delegation/1 and
+  # returns an error if such a delegation is found.
+  defp resolve_conflicts(delegations_of_delegator, delegate_id, _proposal_url = nil) do
+    delegations_of_delegator
+    |> Stream.filter(fn d ->
+      d.delegate_id == delegate_id and d.proposal_url != nil
+    end)
+    |> Enum.each(fn d ->
+      delete_delegation!(d)
+    end)
+
+    {:ok, delegations_of_delegator}
+  end
 
-  defp where_proposal(query, proposal_url),
-    do: query |> where([d], d.proposal_url == ^proposal_url)
+  defp resolve_conflicts(delegations_of_delegator, delegate_id, _proposal_url) do
+    delegations_of_delegator
+    |> Enum.filter(fn d ->
+      d.proposal_url == nil and d.delegate_id == delegate_id
+    end)
+    |> case do
+      [] ->
+        {:ok, delegations_of_delegator}
+
+      [_global_delegation_same_participants] ->
+        {:error,
+         %{
+           message: "Could not create delegation.",
+           details: "A global delegation for the same participants already exists."
+         }}
+    end
+  end
+
+  # Searches the list of delegations for the specified delegator for delegations
+  # of the same type (global or for the same specific proposal) as in the
+  # attributes passed into upsert_delegation/1.
```

Comment:
Looks for a participant's existing delegation of matching type (global or proposal-specific)

Reply:
How about: 'Looks for a _delegator's_ existing delegation of matching type (global or proposal-specific)'

Reply:
That works too. I'm just trying to have less repetition of "delegat*" as it makes both comments and code very hard to reason about, because the words "delegator", "delegation" and "delegate" are so similar.

Reply:
Agree. It's good to get feedback, as it's hard to see the comments/code with fresh eyes when I know what it does. Though, in a week, I won't remember what it does without help from comments / var names, etc. :wink:

Reply:
Definitely, and I know it can feel like nitpicking when asking for changes to comments, but it really pays off down the line, or in a week like you said :)